### PR TITLE
clientv3: initial txn

### PR DESCRIPTION
**Pull request by xiang90**
/cc @heyitsanthony @gyuho @barakmich @mitake

### CODE REVIEW

File: clientv3/kv.go [hunk 1]
```diff
@@ -183,3 +185,96 @@ func (kv *kv) do(op Op) (*pb.ResponseUnion, error) {
...
+	}
+
+	return txn
+}
+
+func (txn *txn) Then(ops ...Op) *txn {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+
+	if txn.cif {
```
Comment (heyitsanthony) [inline clientv3/kv.go]: `cthen`?

**Comment by gyuho** [comment]
LGTM.

File: clientv3/kv.go (hunk 1 above)
Comment (xiang90) [inline clientv3/kv.go]: oh. right.

File: clientv3/kv.go [hunk 2]
```diff
@@ -183,3 +185,96 @@ func (kv *kv) do(op Op) (*pb.ResponseUnion, error) {
 		kv.remote = pb.NewKVClient(kv.conn)
 	}
 }
+
+type txn struct {
```
Comment (heyitsanthony) [inline clientv3/kv.go]: split out all the txn stuff into `txn.go`?

Comment (xiang90) [inline clientv3/kv.go]: sure.

**Comment by xiang90** [comment]
@heyitsanthony PTAL

File: clientv3/kv.go [hunk 3]
```diff
@@ -183,3 +185,96 @@ func (kv *kv) do(op Op) (*pb.ResponseUnion, error) {
...
+		if err == nil {
+			return (*TxnResponse)(resp), nil
+		}
+
+		newConn, cerr := kv.c.retryConnection(kv.conn, err)
+		if cerr != nil {
+			// TODO: return client lib defined connection error
+			return nil, cerr
+		}
+		kv.conn = newConn
```
Comment (heyitsanthony) [inline clientv3/kv.go]: Will the client support code like:
```go
go kv.Txn().Then(foo).Commit()
go kv.Put(bar)
```
or would the Commit and Put need to be wrapped by a mutex?

Comment (xiang90) [inline clientv3/kv.go]: Good question. I am not sure now. But we probably should make all our objects thread-safe in the future.

Comment (heyitsanthony) [inline clientv3/kv.go]: maybe some TODO comment about a possible data race in that case?

Comment (xiang90) [inline clientv3/kv.go]: added.

File: clientv3/compare.go [hunk 4]
```diff
@@ -14,5 +14,76 @@
...
+const (
+	CompareVersion CompareTarget = iota
+	CompareCreated
+	CompareModified
+	CompareValue
+)
+
+type Cmp pb.Compare
+
+func Compare(key string, t pb.Compare_CompareTarget, result string, v interface{}) Cmp {
```
Comment (heyitsanthony) [inline clientv3/compare.go]: it might be worth having `result` be a `rune` or `byte`, not sure

**Comment by heyitsanthony** [comment]
lgtm

File: clientv3/compare.go (hunk 4 above)
Comment (smarterclayton) [inline clientv3/compare.go]: It would be nice to have string and byte operations both available to avoid extra allocations during conversion.

**Pull request merged**